src/main/java/org/zephyrsoft/sdb2/gui/MainWindow.java

```diff
@@ -621,10 +623,113 @@ private void indexAllSongs() {
 		indexer.index(IndexType.ALL_SONGS, songsModel.getSongs());
 	}
 	
+	@Override
+	public void OnIndexChange() {
+		comboBoxLanguage.removeAllItems();
+		Set<String> languages = new HashSet<>();
+		// Fill in used languages:
+		for (Song song : songsModel.getSongs()) {
+			if (song.getLanguage() != null && !song.getLanguage().isBlank()) {
+				languages.add(song.getLanguage());
+			}
+		}
+		for (String language : languages) {
+			comboBoxLanguage.addItem(language);
+		}
+		
+		SwingUtilities.invokeLater(() -> {
+			LOG.debug("Loading songsmodel changes into ui..");
+			
+			Song newSelectedSong = selectNewSongWithUUID != null ? songsModel.getByUUID(selectNewSongWithUUID) : null;
+			if (newSelectedSong != null) {
+				selectNewSongWithUUID = null;
+				// Open song, if just added with handleNewSong
+				setSelectedSong(newSelectedSong, false);
+			} else if (selectedSong != null) {
+				// If current opened song differs, reopen it.
+				// This means overwrite current editor changes without saving them.
+				newSelectedSong = songsModel.getByUUID(selectedSong.getUUID());
+				if (!selectedSong.equals(newSelectedSong)) {
+					if (newSelectedSong == null) {
+						// Song is deleted:
+						setSelectedSong(null, false);
+					} else {
+						// Song is updated
+						newSelectedSong = rebaseChanges(newSelectedSong);
+						setSelectedSong(newSelectedSong, false);
+					}
+				}
+			}
+			
+			// Update songlist / make it visible:
+			if (songsListFiltered != null && selectedSong != null && textFieldFilter != null && !textFieldFilter.getText().isBlank()
+				&& !songsListFiltered.contains(selectedSong)) {
+				textFieldFilter.setText("");
+			} else {
+				applyFilter();
+			}
+		});
+	}
+	
+	private Song rebaseChanges(Song songFromRemote) {
+		if (!isSongDataChanged()) {
+			return songFromRemote;
+		}
+		
+		Song songFromGui = songFromGUI();
+		if (songFromGui.equals(songFromRemote)) {
+			return songFromRemote;
+		}
+		
+		Song localPatch = PatchController.patch(songFromGui, selectedSong);
+		if (localPatch.isEmpty()) {
+			return songFromRemote;
+		}
+		
+		// Merge/Remove equal patches:
+		Song remotePatch = PatchController.patch(songFromRemote, selectedSong);
+		Song mergedPatch = PatchController.mergePatches(remotePatch, localPatch);
+		return PatchController.applyPatch(selectedSong, mergedPatch);
+		
+		// Or simply apply our local changes ontop of the remote changes:
+		// return PatchController.applyPatch(songFromRemote, patch);
```

Comment:
If this is not relevant anymore, please delete it. Otherwise, why is it commented out?

Reply:
Deleted.